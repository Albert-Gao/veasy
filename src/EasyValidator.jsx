import PropTypes from 'prop-types';
/* eslint-disable react/forbid-prop-types,import/no-extraneous-dependencies */
import React from 'react';
import * as lib from './helpers';

export default class EasyValidator extends React.Component {
  getChildren = () => {
    const { schema, formStatus, children } = this.props;
    const names = Object.keys(schema);
    const { fields } = formStatus;
    const newChild = React.Children.map(children, child => {
      const childName = child.props.name;
      if (names.includes(childName)) {
        return React.cloneElement(child, {
          status: fields[childName].status,
          hint: fields[childName].errorText,
          value: fields[childName].value
        });
      }
      return child;
    });
    return newChild;
  };

  handleOnChange = e => (this.validate(e.target));

  validate(target) {
    const { schema, formStatus, update } = this.props;
    lib.startValidating(target, schema, formStatus, update)
  }

  render() {
    const newChildren = this.getChildren();
    return <section onChange={this.handleOnChange}>{newChildren}</section>;
  }
}

EasyValidator.propTypes = {
  schema: PropTypes.object.isRequired,
  formStatus: PropTypes.object.isRequired,
  update: PropTypes.func.isRequired,
  children: PropTypes.any.isRequired
};
