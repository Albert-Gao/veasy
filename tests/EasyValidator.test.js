/* eslint-disable import/no-extraneous-dependencies,react/jsx-filename-extension */
import Enzyme, {mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import React from 'react';
import EasyValidator from '../src/EasyV';
import toJson from 'enzyme-to-json';

Enzyme.configure({ adapter: new Adapter() });

describe('Test the <EasyVLib />', () => {
  let mockSchema;
  let mockTarget;
  let mockComponent;
  const mockSetState = jest.fn();

  beforeEach(() => {
    mockSchema = {
      title: {
        string: {
          minLength: 1,
          default: ''
        }
      }
    };
    mockTarget = {
      name: 'title',
      value: 'abc'
    };
    mockComponent = {
      state: {
        formStatus: {
          isFormOK: false,
          fields: {
            title: {
              status: 'normal',
              errorText: '',
              value: ''
            }
          }
        }
      },
      setState: mockSetState
    };
  });

  afterEach(() => {
    mockSetState.mockReset();
  });

  test('should allows to set 3 props', () => {
    const wrapper = mount(
      <EasyValidator
        schema={mockSchema}
        formStatus={mockComponent.state.formStatus}
        update={mockComponent.setState}
      >
        <input name="title" />
      </EasyValidator>
    );
    expect(wrapper.find(EasyValidator)).toHaveLength(1);
    expect(wrapper.props().schema).toEqual(mockSchema);
    expect(wrapper.props().formStatus).toEqual(mockComponent.state.formStatus);
    expect(wrapper.props().update).toEqual(mockComponent.setState);
  });

  test('Should render a section as container with a onChange function', () => {
    const wrapper = mount(
      <EasyValidator
        schema={mockSchema}
        formStatus={mockComponent.state.formStatus}
        update={mockComponent.setState}
      >
        <input name="title" />
        <input />
        <input />
      </EasyValidator>
    );
    expect(wrapper.find('section')).toHaveLength(1);
    expect(typeof wrapper.find('section').prop('onChange')).toEqual('function');
    wrapper.find('section').prop('onChange')({ target: mockTarget });
  });

  test('Should render 3 inputs as children of section', () => {
    const wrapper = shallow(
      <EasyValidator
        schema={mockSchema}
        formStatus={mockComponent.state.formStatus}
        update={mockComponent.setState}
      >
        <input />
        <input />
        <input />
      </EasyValidator>
    );
    const children = wrapper.find('section').children();
    expect(children).toHaveLength(3);
    expect(children.find('input')).toHaveLength(3);
  });

  test('The 1st children should have extra props', () => {
    const wrapper = shallow(
      <EasyValidator
        schema={mockSchema}
        formStatus={mockComponent.state.formStatus}
        update={mockComponent.setState}
      >
        <input name="title" />
        <input />
        <input />
      </EasyValidator>
    );
    const targetInput = wrapper.find('input').at(0);
    expect(targetInput.prop('name')).toEqual('title');
    expect(targetInput.prop('status')).toEqual('normal');
    expect(targetInput.prop('hint')).toEqual('');
    expect(targetInput.prop('value')).toEqual('');
  });

  it('snapshot tests.', () => {
    const wrapper = mount(
      <EasyValidator
        schema={mockSchema}
        formStatus={mockComponent.state.formStatus}
        update={mockComponent.setState}
      >
        <input name="title" />
        <input />
        <input />
      </EasyValidator>
    );
    // const tree = renderer.create(wrapper).toJSON();
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('snapshot tests using 3rd.', () => {
    const wrapper = mount(
      <EasyValidator
        schema={mockSchema}
        formStatus={mockComponent.state.formStatus}
        update={mockComponent.setState}
      >
        <input name="title" />
        <input />
        <input />
      </EasyValidator>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});