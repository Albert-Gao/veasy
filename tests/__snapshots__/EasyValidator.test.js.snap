// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the <EasyVLib /> snapshot tests using 3rd. 1`] = `
<EasyV
  allState={
    Object {
      "isFormOK": false,
      "title": Object {
        "errorText": "",
        "status": "normal",
        "value": "",
      },
    }
  }
  schema={
    Object {
      "title": Object {
        "string": Object {
          "default": "",
          "minLength": 1,
        },
      },
    }
  }
  update={[Function]}
>
  <section
    onChange={[Function]}
  >
    <input
      hint=""
      key=".0"
      name="title"
      status="normal"
      value=""
    />
    <input
      key=".1"
    />
    <input
      key=".2"
    />
  </section>
</EasyV>
`;

exports[`Test the <EasyVLib /> snapshot tests. 1`] = `
"<EasyV schema={{...}} allState={{...}} update={[Function]}>
  <section onChange={[Function]}>
    <input name=\\"title\\" status=\\"normal\\" hint=\\"\\" value=\\"\\" />
    <input />
    <input />
  </section>
</EasyV>"
`;
